### PR TITLE
Feature review

### PR DESCRIPTION
**Pull request by MihaiStanusoiu**

### CODE REVIEW

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 1]
```diff
@@ -32,6 +32,11 @@
     private ConsumerRepository repoConsumer;
 
 
+    @RequestMapping(value = "/get/{id}", method = RequestMethod.GET)
+    public @ResponseBody JobEntity get(@PathVariable("id") final Integer id) {
+        return repo.findOne(id);
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: Write better names (idk if it's you or Nico, but "repo" does not say "jobRepository" or "repositoryJob").
It's hard to follow code in reviews, when you don't how exactly what the variable represents. (because some of the lines from the file are omitted).

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 2]
```diff
@@ -54,4 +59,24 @@
         }
     }
 
+
+    @RequestMapping(value = "/reviews/provider/{id}/{nr}", method = RequestMethod.GET)
+    public @ResponseBody List<ReviewEntity> getTopReviewsForProvider(@PathVariable("id") final Integer id, @PathVariable("nr") final Integer nr) {
+        ProviderEntity provider = repoProv.findOne(id);
+        List<ReviewEntity> reviews = new ArrayList<>(provider.getReviews());
+        reviews.sort((r1, r2) -> r2.getRating() - r1.getRating());
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: Better Names: r1 = left, r2 = right (or lhs, rhs) also you can sort based on field (like rating) you don't have to tell it how to sort, just the base.

Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: Use what @nicofecioru  worked on (annotations that do the sorting for you)! Don't sort in the RestController, it's better if the db sorts and returns the first N entities directly to you. (as performance goes)

PS. you can sort directly on the getReviews() <-- because the list you're mutating it's not saved in the database ones the process is done.  (but do that I told you ... use what @nicofecioru  created with @decebaldecebal )

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 3]
```diff
@@ -54,4 +59,24 @@
...
+
+        return reviews.subList(0, nr);
+    }
+
+    @RequestMapping(value = "/reviews/consumer/{id}/{nr}", method = RequestMethod.GET)
+    public @ResponseBody List<ReviewEntity> getTopReviewsForConsumer(@PathVariable("id") final Integer id, @PathVariable("nr") final Integer nr) {
+        ConsumerEntity consumer = repoConsumer.findOne(id);
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: Same as above, the sort and filter process should start in db, you don't need to take the whole data from database. ( @decebaldecebal will tell you the same -- I will add him as a reviewer as well)

File: src/main/java/meeseeks/box/controller/JobController.java (hunk 1 above)
Comment (decebaldecebal) [reply src/main/java/meeseeks/box/controller/JobController.java]: I think "jobRepo" works best, "repository" is too long for a suffix.

**Comment by decebaldecebal** [comment]
Please take a look at the latest changes in the JobRepository and how to limit the results using queries.

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 4]
```diff
@@ -89,24 +91,21 @@ private ConsumerEntity getConsumerById(final Integer id) {
...
-    public @ResponseBody List<ReviewEntity> getTopReviewsForConsumer(@PathVariable("id") final Integer id, @PathVariable("nr") final Integer nr) {
-        ConsumerEntity consumer = repoConsumer.findOne(id);
-        List<ReviewEntity> reviews = new ArrayList<>(consumer.getReviews());
-        reviews.sort((r1, r2) -> r2.getRating() - r1.getRating());
-
-        return reviews.subList(0, nr);
+    @RequestMapping(value = "/consumer/reviews/{id}/{limit}", method = RequestMethod.GET)
+    public @ResponseBody List<ReviewEntity> getTopReviewsForConsumer(@PathVariable("id") final Integer id, @PathVariable("limit") final Integer limit) {
+        return consumerRepository.findTopReviewsForConsumer(
+                consumerRepository.findOne(id),
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: Don't use "findOne" use findById (from BaseCrudRepository) --> because returns an Optional (so you don't have to consider the null case)

findOne might return you a null references => PointerNullException

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 5]
```diff
@@ -89,24 +91,21 @@ private ConsumerEntity getConsumerById(final Integer id) {
...
 
-    @RequestMapping(value = "/reviews/consumer/{id}/{nr}", method = RequestMethod.GET)
-    public @ResponseBody List<ReviewEntity> getTopReviewsForConsumer(@PathVariable("id") final Integer id, @PathVariable("nr") final Integer nr) {
-        ConsumerEntity consumer = repoConsumer.findOne(id);
-        List<ReviewEntity> reviews = new ArrayList<>(consumer.getReviews());
-        reviews.sort((r1, r2) -> r2.getRating() - r1.getRating());
-
-        return reviews.subList(0, nr);
+    @RequestMapping(value = "/consumer/reviews/{id}/{limit}", method = RequestMethod.GET)
+    public @ResponseBody List<ReviewEntity> getTopReviewsForConsumer(@PathVariable("id") final Integer id, @PathVariable("limit") final Integer limit) {
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/controller/JobController.java]: The order of the reviews are based on date (we need the latest reviews not the best)!

Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/JobController.java]: "Extend Job RestController with a function that gets the latest N reviews for consumer."<-- your job

File: src/main/java/meeseeks/box/repository/ConsumerRepository.java [hunk 6]
```diff
@@ -1,12 +1,24 @@
...
  * @author Alexandru Stoica
  * @version 1.0
  */
 
 @Repository
-public interface ConsumerRepository extends BaseCrudRepository<ConsumerEntity, Integer> { }
+public interface ConsumerRepository extends BaseCrudRepository<ConsumerEntity, Integer> {
+
+    @Query("select r from ConsumerEntity c join c.reviews r where c = ?1 order by r.rating desc")
+    List<ReviewEntity> findTopReviewsForConsumer(final @NotNull ConsumerEntity consumer, Pageable pageable);
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/repository/ConsumerRepository.java]: Same here! We need the latest, not the best (aka by rating)!
You should name the function findReviewsByRatingForConsumer

File: src/main/java/meeseeks/box/repository/ProviderRepository.java [hunk 7]
```diff
@@ -1,13 +1,24 @@
...
  * @author Alexandru Stoica
  * @version 1.0
  */
 
 @Repository
-public interface ProviderRepository extends BaseCrudRepository<ProviderEntity, Integer> { }
+public interface ProviderRepository extends BaseCrudRepository<ProviderEntity, Integer> {
+
+    @Query("select r from ProviderEntity p join p.reviews r where p = ?1 order by r.rating desc")
+    List<ReviewEntity> findTopReviewsForProvider(final @NotNull ProviderEntity provider, Pageable pageable);
```
Comment (alexandrustoica) [inline src/main/java/meeseeks/box/repository/ProviderRepository.java]: Same as before + you should move this in the ReviewRepository

**Comment by decebaldecebal** [comment]
I do not know what you have done but your pull request contains too many changes and I can not review them since I see changes that I or other people have made.Please update your branch like this:

Make sure you do not have any changed files then do:
git checkout develop
git pull

Make sure that the pull is successful

git checkout - (to get back to your branch)
git rebase develop (this should rebase your changes on develop, resolve any conflicts that appear by following the instructions in the command line)

After the rebase is completed do:

git push --force (to update your branch with all the changes)

This way the Files Changed tab will show only what you have changed compared to develop and not all the commit on develop.
